feat: add updateFarmVest init function

amusingaxl opened the pull request.

amusingaxl committed:
feat: add lsSKY -> Sky farm deployment script

amusingaxl committed:
refactor: add full init library for LSSKY -> SKY farm

amusingaxl committed:
test: add coverage for LsskySkyFarmingInit

amusingaxl committed:
refactor: generalize farm initialization library

amusingaxl committed:
refactor: rename parameters in TreasuryFundedFarmingInit test

amusingaxl committed:
feat: add regular farming init

amusingaxl committed:
refactor: improve testing for initializatino of farms

amusingaxl committed:
refactor: remove admin from parameters

amusingaxl committed:
chore: improve formatting

amusingaxl committed:
chore: simplify formatting and linting

amusingaxl committed:
chore: remove TODO

amusingaxl committed:
test: add integration tests for farms after initialization

amusingaxl committed:
feat: add updateFarmVest init function

amusingaxl committed:
chore: remove unused structs

amusingaxl committed:
fix: wrong chainlog key in deployment script

amusingaxl committed:
refactor: explicitly name and check for `DssVestTransferrable`

amusingaxl committed:
chore: clean up interfaces

amusingaxl committed:
fix: typo in function name

amusingaxl committed:
refactor: add comments to FarmingInitParams fields

amusingaxl committed:
fix: make CI use a stable Foundry version

amusingaxl committed:
fix: add ETH_RPC_URL to CI env

amusingaxl committed:
chore: fix typos

amusingaxl committed:
test: add coverage for updating `cap`

amusingaxl committed:
test: add fuzzing for e2e tests

amusingaxl committed:
Merge branch 'feat/sky-farms-deployment' into feat/update-farm-vest

amusingaxl committed:
refactor: make dependency on `DssVestTransferrable` explicit

amusingaxl committed:
fix: wrong interface name after merge

amusingaxl committed:
fix: missing interface method

amusingaxl committed:
fix: missing interface method (again)

amusingaxl committed:
Merge branch 'master' into feat/update-farm-vest

Signed-off-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>

amusingaxl committed:
refactor: remove noise from implementation

amusingaxl committed:
refactor: add test coverage for `updateVestFarm`

0xLaz3r commented on script/dependencies/treasury-funded-farms/TreasuryFundedFarmingInit.sol:
```diff
@@ -117,14 +124,64 @@ library TreasuryFundedFarmingInit {
+        require(vest.czar() == address(this), "initFarm/vest-czar-mismatch");
+
+        address rewardsToken = dist.gem();
+        ERC20Like rewardsTokenContract = ERC20Like(rewardsToken);
```
The local variable might not be needed
```suggestion
        ERC20Like rewardsTokenContract = ERC20Like(dist.gem());
```

amusingaxl replied:
Updated the code.

0xLaz3r commented on script/dependencies/treasury-funded-farms/TreasuryFundedFarmingInit.sol:
```diff
@@ -117,14 +124,64 @@ library TreasuryFundedFarmingInit {
+            dist.distribute();
+        }
+
+        // Get the remaining allawance of the previous vesting stream.
```
Typo
```suggestion
        // Get the remaining allowance of the previous vesting stream.
```

amusingaxl replied:
Thanks for spotting. Fixed.
Just noticed that my editor font's italic style has very similar `o` and `a`.

0xLaz3r commented on script/dependencies/treasury-funded-farms/TreasuryFundedFarmingInit.sol:
```diff
@@ -117,14 +124,64 @@ library TreasuryFundedFarmingInit {
+        uint256 prevVestRxd = vest.rxd(prevVestId);
+
+        // Adjust the allowance considering the previous vest unclaimed amount and the new vest total.
+        rewardsTokenContract.approve(address(vest), currAllowance + p.vestTot - (prevVestTot - prevVestRxd));
```
Isn't `prevVestTot - prevVestRxd` equal to `currAllowance` or could there be any potential discrepancy?

amusingaxl replied:
Allowance is cumulative to the many streams that might exist in the contract.
`tot` and `rxd` are per stream.
`currAllowance == prevVestTot - prevVestRx` iff there's only one stream in the contract.

amusingaxl committed:
refactor: address review comments

oddaf commented on script/dependencies/treasury-funded-farms/TreasuryFundedFarmingInit.sol:
```diff
@@ -117,14 +124,63 @@ library TreasuryFundedFarmingInit {
         LockstakeEngineLike(lockstakeEngine).addFarm(p.rewards);
     }
+
+    function updateFarmVest(FarmingUpdateVestParams memory p) internal returns (FarmingUpdateVestResult memory r) {
```
Shouldn't we also update the cap here? If the rate ends up being larger than cap, the stream will be rate limited. In line 90 the `initFarm` function increases it if necessary (another option is to require a param for it).

oddaf replied:
Some of the checks performed in initFarm could be useful here:

- That p.dist is a valid VestedRewardsDistributionLike contract
- That p.vestTot > 0
- That p.vestTau > 0 (division by zero risk in cap calculation if added)

oddaf replied:
Should the chainlog be updated for `dist`?

amusingaxl replied:
Why would the chainlog require to be updated? We're not changing the distribution contract, only it's vesting stream.

amusingaxl replied:
> Some of the checks performed in initFarm could be useful here:
> 
>     * That p.dist is a valid VestedRewardsDistributionLike contract
> 
>     * That p.vestTot > 0
> 
>     * That p.vestTau > 0 (division by zero risk in cap calculation if added)

Done in e93630c

amusingaxl replied:
> Shouldn't we also update the cap here? If the rate ends up being larger than cap, the stream will be rate limited. In line 90 the `initFarm` function increases it if necessary (another option is to require a param for it).

Great catch!
Done in ad40dee

oddaf commented on script/dependencies/treasury-funded-farms/TreasuryFundedFarmingInit.sol:
```diff
@@ -117,14 +124,63 @@ library TreasuryFundedFarmingInit {
+        VestedRewardsDistributionLike dist = VestedRewardsDistributionLike(p.dist);
+        DssVestTransferrableLike vest = DssVestTransferrableLike(dist.dssVest());
+        // Note: `vest` is expected to be of type `DssVestTransferrable`
+        require(vest.czar() == address(this), "initFarm/vest-czar-mismatch");
```
```suggestion
        require(vest.czar() == address(this), "updateFarmVest/vest-czar-mismatch");
```

amusingaxl replied:
Fixed in e93630c

oddaf commented on script/dependencies/treasury-funded-farms/TreasuryFundedFarmingInit.sol:
```diff
@@ -41,48 +39,56 @@ struct FarmingInitResult {
+        require(dist.vestId() == 0, "initFarm/dist-vest-id-already-set");
+        require(dist.stakingRewards() == p.rewards, "initFarm/dist-staking-rewards-mismatch");
 
         // Set `dist` with  `rewardsDistribution` role in `rewards`.
```
```suggestion
        // Set `dist` with `rewardsDistribution` role in `rewards`.
```

double space

amusingaxl replied:
Fixed in 9e6e4cb.

amusingaxl committed:
refactor: add extended sanity checks to updateFarmVest

- Validate vestTot > 0 to prevent zero total vesting
- Validate vestTau > 0 to prevent division by zero
- Check that distribution vestId is already set (not zero)
- Verify vest gem matches distribution gem
- Fix error message prefix from initFarm to updateFarmVest
- Add comprehensive test coverage for invalid parameters

amusingaxl committed:
refactor: add cap adjustment handling for updateFarmVest

- Check if vest cap needs adjustment when required rate exceeds current cap
- Set cap to rate with 10% buffer when adjustment is needed
- Refactor allowance calculation into scoped block for clarity
- Add comprehensive test coverage for cap adjustment scenarios
- Add tests for allowance calculation and vest creation parameters

amusingaxl committed:
chore: remove extra whitespace

amusingaxl committed:
fix: move vest parameter assertions to correct test function

amusingaxl committed:
refactor: separate integration tests into dedicated file

amusingaxl committed:
test: improve assertion error messages in integration tests

Replace generic error messages with descriptive ones that clearly explain
what operation failed and what the expected behavior should be.

0xLaz3r reviewed (approved):
LGTM!
